src/prg_quantumdynamics_mod.F90

```diff
@@ -0,0 +1,444 @@
+  prg_allocate_bml, prg_deallocate_bml
+
+contains
+
+  !> Provides perturbation to initial density matrix in the form of an electric
+  !! field kick.
+  !! This routine does:
+  !! \f$\hat{\rho_{kick}} =
+  !! \exp{\frac{-i}{\hbar}\hat{V}}\hat{\rho}\hat{S}\exp{\frac{i}{\hbar}\hat{V}}\hat{S^{-1}}\f$
+  !! where \f$\hat{V}\f$ is the field disturbance.
+  !! \param kick_direc the direction of the kick in the electric field
+  !! \param kick_mag the magnitude of the kick in the electric field
+  !! \param dens the initial density matrix to be kicked.
+  !! \param N the number of orbitals in the density matrix
+  !! \param S the overlap matrix
+  !! \param SINV the inverse of the overlap matrix
+  !! \param which_atom vector containing atom identification
+  !! \param r direction vector for kick based on atom and kick_direc
+  !!
+  subroutine prg_kick_density(kick_direc,kick_mag,dens,N,S,SINV,&
+             which_atom,r,bmltype)
+    implicit none
+    integer                                     :: i
+    integer, intent(in)                         :: kick_direc, N
+    integer, allocatable, intent(in)            :: which_atom(:)
+    real(dp), allocatable                       :: r(:,:)
+    real(dp)                                    :: kick_mag, thresh
+    complex(dp), allocatable, intent(inout)     :: dens(:,:)
+    complex(dp), allocatable                    :: tmat1(:,:), tmat2(:,:), S(:,:), SINV(:,:)
+    complex(dp)                                 :: telem
+    type(bml_matrix_t)                          :: T1, T2, RHO, OVER, OINV
+    character(len=*), intent(in)                :: bmltype
+
+    thresh=1.0d-5
+    allocate (tmat1(N,N))
+    allocate (tmat2(N,N))
+
+    tmat1 = 0.0_dp
+    tmat2 = 0.0_dp
+
+    do i=1,N
+       telem = cmplx(0.0_dp,-kick_mag*r(which_atom(i),kick_direc))
+       tmat1(i,i) = exp( telem)
+       tmat2(i,i) = exp( -telem)
+    enddo
+
+    call bml_zero_matrix(bmltype, BML_ELEMENT_COMPLEX,dp,N,N, T1)
+    call bml_zero_matrix(bmltype, BML_ELEMENT_COMPLEX,dp,N,N, T2)
+    call bml_zero_matrix(bmltype, BML_ELEMENT_COMPLEX,dp,N,N, RHO)
+    call bml_zero_matrix(bmltype, BML_ELEMENT_COMPLEX,dp,N,N, OVER)
+    call bml_zero_matrix(bmltype, BML_ELEMENT_COMPLEX,dp,N,N, OINV)
+
+    call bml_convert_from_dense(bmltype,tmat1,T1,thresh,N)
+    call bml_convert_from_dense(bmltype,tmat2,T2,thresh,N)
+    call bml_convert_from_dense(bmltype,dens,RHO,thresh,N)
+    call bml_convert_from_dense(bmltype,S,OVER,thresh,N)
+    call bml_convert_from_dense(bmltype,SINV,OINV,thresh,N)
+
+    call bml_multiply(T1,RHO,T2)
+    call bml_multiply(T2,OVER,RHO)
+    call bml_convert_from_dense(bmltype,tmat2,T1,thresh,N)
+
+    call bml_multiply(RHO,T1,T2)
+    call bml_multiply(T2,OINV,RHO)
+
+    write(*,*)"#######Density matrix Sparsity######"
+    write(*,*)"thr,sparsity ",thresh,bml_get_sparsity(RHO,thresh)
+
+    call bml_convert_to_dense(RHO,dens)
+
+    call bml_deallocate(T1)
```

Comment:
Can we allocate and deallocate right before and after the matrix is used? This is just because the matrices are large and the less we have in memory the better